README.markdown
# BELGIAN 2050 CALCULATOR TOOL

A C version and ruby wrapper for the Belgian 2050 calcualtor

## GOTCHAS

Some versions have a special formula in 2050!B2 that the translator doesn't recognise. Just write 2050 in that cell and recompile.

Some tests fail for columns AN and AM on OUTPUT. I think this is due to rounding differences between excel and C.

## DEPENDENCIES

1. ruby 1.9.2 (including development headers)
2. basic c development headers

This has ONLY been tested on OSX and on Ubuntu 64 bit EC2 ami.
Grateful for reports from other platforms. 

In the util folder there is an example script that creates a new EC2 EMI, installs all the dependencies and then compiles the gem. It may be useful if you are trying to figure out the complete set of dependencies.

## INSTALLATION

Note that this compiles the underlying c code, which might take 10-20 minutes or so

    gem install belgium_2050_model
  
## UPDATING TO NEWER VERSIONS OF EXCEL MODEL

First of all, you need to be working on the github version of the code, not the rubygem:
    
    git clone http://github.com/decc/belgium_2050_model

Then put the new spreadsheet in spreadsheet/2050Model.xlsx

Then, from the top directory of the gem:
  
    bundle
    bundle exec rake
  
The next step is to check whether lib/belgium_2050_model/belgium_2050_model_result.rb and lib/belgium_2050_model/model_structure.rb need to be altered so that they
pick up the correct places in the underlying excel.
  
The final stage is to build and install the new gem:
    
    gem build belgium_2050_model.gemspec
    gem install belgium_2050_model-<version>.gem 

... where <version> is the version number of the gem file that was created in the folder.
  
Now follow the instructions in the twenty-fifty server directory in order to ensure that it is using this new version of the gem.


CHANGES
# Head

# 0.0.7

Updated model to v82

# 0.0.6

Updated model to v80

# 0.0.5

Updated model to v71

# 0.0.4

Updated model

# 0.0.3

Tweaks to model, tweaks to results.

# 0.0.2

Starting to return useful results

# 0.0.1

First version of the model that compiles without significant error


test/test_belgium_2050_model_result.rb
# coding: utf-8
# Test for belgium_2050_model_result
require 'rubygems'
gem 'minitest'
require 'test/unit'
require_relative '../lib/belgium_2050_model'
require 'pp'

class TestBelgium2050ModelResult < Test::Unit::TestCase

  def test_results
    @model = Belgium2050ModelResult.new
    tests_folder = File.join(File.dirname(__FILE__),"expected_results")
    Dir[File.join(tests_folder, "*.json")].each do |test|
      expected = IO.readlines(test)
      pathway = expected.shift
      expected = expected.join
      puts "Testing #{pathway}"
      @model.reset
      result = @model.calculate_pathway(pathway).pretty_inspect
      if result != expected
        File.open(test+".actual",'w') do |f|
          f.puts pathway
          f.puts result
        end
      end
      assert_equal expected, result
    end
  end
end

    


test/create_belgium_2050_model_result.rb
# coding: utf-8
# Create a test for belgium_2050_model_result
require_relative '../lib/belgium_2050_model'
require 'pp'

pathway = ARGV[0].strip
tests_folder = File.join(File.dirname(__FILE__),"expected_results")

puts "Creating a copy of the output for pathway #{pathway} and puting it in #{tests_folder}"
puts "To compare future versions of the model with this test result, run test_belgium_2050_model_result.rb"

model = Belgium2050ModelResult.new
File.open(File.join(tests_folder,"#{pathway}.json"),'w') do |f|
  result = model.calculate_pathway(pathway).pretty_inspect
  f.puts pathway
  f.puts result
end


util/build-ubuntu-native-binary-gem.rb
# This file is used to create the ubuntu native binary
# version of the ruby gem. It does so by setting up
# an aws instance running ubuntu, building the gem there
# and then downloading the resulting gem.

# To use:
# ruby build-ubuntu-native-binary-gem.rb [branch]
# 
# If you don't specify a branch it will compile the gem in 
# the master branch of http://github.com/belgium/belgium_2050_model.git 

branch = ARGV[0] || "master"
server_name = ARGV[1] || "belgium_2050_model_gem_build_#{branch}"

require 'fog'
require 'pry'

# See https://github.com/fog/fog/ for details on
# how to edit this file

# Need to have a ~/.fog file with AWS credentials
# Sign up with aws.amazon.com then create a ~/.fog
# file with this sort of thing:
# :default:
#   :provider: AWS
#   :aws_secret_access_key: <GET THIS FROM AMAZON>
#   :aws_access_key_id: <GET THIS FROM AMAZON>
#   :region: eu-west-1
puts "Connecting to AWS"
connection =  Fog::Compute.new({
  :provider => 'AWS',
})

servers = connection.servers.all('tag:Name' => server_name)
# FIXME: Needs to check if existing servers are terminated or available!
if servers.empty?
  # This sets up an Ubuntu 12.0.4 Precise high compute instance 
  # As of 22/July/2012 this costs $0.744 per Hour
  puts "Bootstrapping a server"
  server = connection.servers.bootstrap(
    #:image_id => 'ami-1de8d369', # Ubuntu 12.04
    :image_id => 'ami-41cacb35', # Ubuntu 11.10
    #:image_id => 'ami-bbadb0cf', # Ubuntu 10.04 eu-west-1 64-bit instance storage
    :flavor_id=> 'c1.xlarge',
    :private_key_path => '~/.ssh/id_rsa', # Change this if you use something else
    :public_key_path => '~/.ssh/id_rsa.pub', # Change this if you use something else
    :username => 'ubuntu',
    :tags => { 'Name' => server_name }
  )
else
  server = servers.first
  puts "Retrieving server #{server.dns_name}"
end

puts "Waiting for server to be ready"
server.wait_for { ready? }
puts "Sever ready"

puts "You can now:"
puts "ssh ubuntu@#{server.dns_name}"

commands = <<EOT
## REQUIRED LIBRARIES

# Basics
sudo apt-get update
sudo apt-get install -y git build-essential libxml2-dev libxslt-dev zip unzip
sudo apt-get install -y libcurl4-openssl-dev libssl-dev 

# Ruby
sudo apt-add-repository -y ppa:brightbox/ruby-ng-experimental
sudo apt-get update 
sudo apt-get install -y ruby1.9.3

## THE CODE
git clone http://github.com/decc/belgium_2050_model.git

# Bundler
sudo gem install --no-ri --no-rdoc bundler

# Compile and build
pwd
cd /home/ubuntu/belgium_2050_model
pwd
git checkout #{branch}
bundle
cd ext
rake
cd ..
gem build belgium_2050_model.gemspec.binary
EOT

require 'tempfile'
script = Tempfile.new 'setup'
script.puts commands
script.close

puts "Uploading setup code"
server.scp_upload(script.path,'setup.sh')

puts "Running setup code"
results = server.ssh('sh setup.sh')
results.each { |r| r.display_stdout }
results.each { |r| r.display_stderr }

puts "Server setup"
gem_name = server.ssh("ls belgium_2050_model | grep .gem$").first.stdout.strip

puts "Tyring to download gem #{gem_name}"
server.scp_download("/home/ubuntu/belgium_2050_model/#{gem_name}",".")
puts "Downloaded"

puts "Tearing down server"
server.destroy



lib/belgium_2050_model/model_structure.rb
require_relative '../belgium_2050_model'
require 'singleton'

class ModelStructure < Belgium2050ModelUtilities
  include Singleton
  
  attr_accessor :excel, :choices
    
  def initialize
    @excel = Belgium2050ModelShim.new
    @choices = []
    types.each_with_index do |choice_type,i|
      case choice_type
      when nil, 0.0; next
      when /[abcd]/i; choices << ModelAlternative.new(i,names[i],choice_type,descriptions[i],long_descriptions[i])
      else; choices << ModelChoice.new(i,names[i],choice_type,descriptions[i],long_descriptions[i])
      end
    end
  end
  
  def reported_calculator_version
    excel.control_m1
  end
  
  def controls(col1, col2)
    (4..55).to_a.map { |row| excel.send("control_#{col1}#{row}") } + (4..32).to_a.map { |row| excel.send("control_#{col2}#{row}") }
  end

  def types
    @types ||= controls('p', 'an')
  end
  
  def names
    @names ||= controls('f', 'ad')
  end

  def descriptions
    @descriptions ||= (4..55).to_a.map { |row| [r("control_q#{row}"),r("control_r#{row}"),r("control_s#{row}"),r("control_t#{row}")] } + (4..32).to_a.map { |row| [r("control_ao#{row}"),r("control_ap#{row}"),r("control_aq#{row}"),r("control_ar#{row}")] }
  end

  def long_descriptions
    @long_descriptions ||= (4..55).to_a.map { |row| [r("control_u#{row}"),r("control_v#{row}"),r("control_w#{row}"),r("control_x#{row}")] } + (4..32).to_a.map { |row| [r("control_as#{row}"),r("control_at#{row}"),r("control_au#{row}"),r("control_av#{row}")] }
  end
    
  def demand_choices
    choices[0..22]
  end

  def industry_choices
    choices[23..43]

  end
  
  def supply_choices
    choices[44..64]
  end

  def example_pathways
    @example_pathways ||= generate_example_pathways
  end
  
  def generate_example_pathways
    %w{az ba bb bc bd be bf bg bh bi}.map do |column|
      code = ((4..55).to_a + (62..90).to_a).map { |row| r("control_#{column}#{row}") }
      {
        name: r("control_#{column}1"),
        code: convert_float_to_letters(code).join,
      }
    end
  end

  # NB: The Belgian calcualtor has some inputs 10..40 rather than 1..4
  def convert_float_to_letters(array)
    array.map.with_index do |entry, i|
      type = types[i]
      entry = 0 if type == nil || type == 0
      entry = (entry / 10.0) if entry && type && type > 4
      case entry
      when nil; 0
      when Float; Belgium2050ModelResult::FLOAT_TO_LETTER_MAP[entry] || entry
      else entry
      end
    end
  end
  
  # FIXME: Only wraps one line into two
  def wrap(string, wrap_at_length = 45)
    return "" unless string
    string = string.to_s
    length_so_far = 0
    string.split.partition do |word| 
      length_so_far = length_so_far + word.length + 1 # +1 for the trailing space 
      length_so_far > wrap_at_length
    end.reverse.map { |a| a.join(" ") }.join("\n")
  end

end


lib/belgium_2050_model/belgium_2050_model_result.rb
require_relative '../belgium_2050_model'

class Belgium2050ModelResult < Belgium2050ModelUtilities  
  attr_accessor :excel, :pathway
  
  def initialize
    @excel = Belgium2050ModelShim.new
  end
  
  def self.calculate_pathway(code)
    new.calculate_pathway(code)
  end
  
  def calculate_pathway(code)
    Thread.exclusive do 
      reset
      @pathway = { _id: code, choices: set_choices(code) }
      # One method per view
      greenhouse_gas_tables
      all_energy_tables
      electricity_tables
      energy_security_tables
      energy_flow_tables
      area_tables
      story_tables
      air_quality_tables # Not implemented
      simple_costs_tables
      complex_costs_tables
    end
    return pathway
  end
  
def greenhouse_gas_tables
    # Each number is a row on the OUTPUT worksheet
    pathway[:ghg] = table 399, 400, 401, 402, 403, 404, 405, 406, 407, 409, 411, 412 
    pathway[:ghg_by_sector] = table 363, 364, 365, 366, 367, 368, 369, 370, 371, 372, 373, 374, 375, 376, 377, 378
end  

  def all_energy_tables
    # Each number is a row on the OUTPUT worksheet
    pathway[:final_energy_demand, ] = table 12, 16, 19, 21
    pathway[:primary_energy_supply] = table 27, 28, 29, 30, 31, 32, 33, 34, 36, 41, 44, 48, 51, 52
  end

  def electricity_tables
    e = {}
    # Each number is a row on the OUTPUT worksheet
    e[:emissions] = table 441, 442, 443, 444
    e[:demand] = table 94, 97, 110, 120
    e[:supply] = table 130, 131, 135, 136, 137, 138, 139, 141, 142, 143, 146, 149, 151
    pathway[:electricity] = e
  end

  def energy_security_tables
    energy_imports
    energy_diversity
  end

  def energy_imports
    sheet_name = "online_graphs_and_color_codes"
    i = {}
    (189..195).each do |row|
      i[r("#{sheet_name}_b#{row}")] = {
        '2010' => {
          :quantity => r("#{sheet_name}_d#{row}").to_f,
          :proportion => r("#{sheet_name}_e#{row}").to_f
        },
        '2050' => {
          :quantity => r("#{sheet_name}_g#{row}").to_f,
          :proportion => r("#{sheet_name}_h#{row}").to_f
        }
      }
    end
    pathway['imports'] = i
  end

  def energy_diversity
    sheet_name = "online_graphs_and_color_codes"
    i = {}
    (202..211).each do |row|
      i[r("#{sheet_name}_b#{row}")] = {
        '2010' => "#{(r("#{sheet_name}_d#{row}").to_f*100).round}%",
        '2050' => "#{(r("#{sheet_name}_g#{row}").to_f*100).round}%"
      }
    end
    pathway['diversity'] = i
  end

  def energy_flow_tables
    s = [] 
    (6..93).each do |row|
      s << [r("flows_c#{row}"),r("flows_m#{row}"),r("flows_d#{row}")]
    end
    pathway[:sankey] = s
  end

  def area_tables
    m = {}
    [6..11,15..18,22..22,27..27,31..36].each do |range|
      range.to_a.each do |row|
        m[r("land_use_c#{row}")] = r("land_use_p#{row}")
      end
    end
    pathway['map'] = m
  end

  def story_tables
    pathway[:ghg][:percent_reduction_from_1990] = ((1-r("output_ak424")) * 100).round
    pathway[:electricity][:capacity] = table 160, 164, 165, 166, 167, 168, 172, 173, 174
    sheet_name = "online_graphs_and_color_codes"
    pathway[:heating] = {}
    h = {}
    (84..88).each do |row|
      h[r("#{sheet_name}_b#{row}")] = r("#{sheet_name}_c#{row}")
    end
    pathway[:heating][:residential] = h
    h = {}
    (99..103).each do |row|
      h[r("#{sheet_name}_b#{row}")] = r("#{sheet_name}_c#{row}")
    end
    pathway[:heating][:commercial] = h
    pathway[:security] = {
      :imports => {
        :'2010' => r("#{sheet_name}_o92"),
        :'2050' => r("#{sheet_name}_p92")
      },
      :import_proportion => {
        :'2010' => r("#{sheet_name}_o93"),
        :'2050' => r("#{sheet_name}_p93")
      },
      :number_of_technologies_to_cover_eighty_five_percent => {
        :'2010' => r("#{sheet_name}_o98"),
        :'2050' => r("#{sheet_name}_p98")
      }
    }
  end

  def air_quality_tables
    # Not implemented
  end

  def simple_costs_tables
    c = {}
    c[:sector] = table 492, 497, 502, 507
    c[:type] = table 509, 510, 511
    pathway[:simple_costs] = c
  end

  def complex_costs_tables
    t = {}
    low_start_row = 3
    point_start_row = 57
    high_start_row = 112
    number_of_components = 49
    
    # Normal cost components
    (0..number_of_components).to_a.each do |i|
            
      name          = r("costpercapita_b#{low_start_row+i}")
      
      low           = r("costpercapita_as#{low_start_row+i}")
      point         = r("costpercapita_as#{point_start_row+i}")
      high          = r("costpercapita_as#{high_start_row+i}")
      range         = high - low
      
      finance_low   = 0 # r("costpercapita_cp{low_start_row+i}") # Bodge for the zero interest rate at low
      finance_point = r("costpercapita_cp#{point_start_row+i}")
      finance_high  = r("costpercapita_cp#{high_start_row+i}")
      finance_range = finance_high - finance_low
      
      costs = {low:low,point:point,high:high,range:range,finance_low:finance_low,finance_point:finance_point,finance_high:finance_high,finance_range:finance_range}
      if t.has_key?(name)
        t[name] = sum(t[name],costs)
      else
        t[name] = costs
      end
    end
    
    # Merge some of the points
    t['Coal'] = sum(t['Indigenous fossil-fuel production - Coal'],t['Balancing imports - Coal'])
    t.delete 'Indigenous fossil-fuel production - Coal'
    t.delete 'Balancing imports - Coal'
    t['Oil'] = sum(t['Indigenous fossil-fuel production - Oil'],t['Balancing imports - Oil'])
    t.delete 'Indigenous fossil-fuel production - Oil'
    t.delete 'Balancing imports - Oil'
    t['Gas'] = sum(t['Indigenous fossil-fuel production - Gas'],t['Balancing imports - Gas'])
    t.delete 'Indigenous fossil-fuel production - Gas'
    t.delete 'Balancing imports - Gas'
    
    # Finance cost
    name          = "Finance cost"
    
    low           = 0 # r("costpercapita_cp#{low_start_row+number_of_components+1}") # Bodge for the zero interest rate at low
    point         = r("costpercapita_cp#{point_start_row+number_of_components+1}")
    high          = r("costpercapita_cp#{high_start_row+number_of_components+1}")
    range         = high - low
    
    finance_low   = 0 # r("costpercapita_cp{low_start_row+i}") # Bodge for the zero interest rate at low
    finance_point = 0
    finance_high  = 0
    finance_range = finance_high - finance_low
    
    t[name] = {low:low,point:point,high:high,range:range,finance_low:finance_low,finance_point:finance_point,finance_high:finance_high,finance_range:finance_range}
  
    pathway['cost_components'] = t
  end
  
  # Helper methods
 
  def table(*rows)
    t = {}
    rows.each do |row|
      t[label("output",row)] = annual_data("output",row)
    end
    t
  end
  
  def label(sheet,row)
    r("#{sheet}_e#{row}").to_s
  end
  
  def annual_data(sheet,row)
    ['ac','ad','ae','af','ag','ah','ai','aj','ak'].map { |c| r("#{sheet}_#{c}#{row}") }
  end
  
  def sum(hash_a,hash_b)
    return nil unless hash_a && hash_b
    summed_hash = {}
    hash_a.each do |key,value|
      summed_hash[key] = value + hash_b[key]
    end
    return summed_hash
  end
  
  # Set the 9 decimal points between 1.1 and 3.9
  FLOAT_TO_LETTER_MAP = Hash["abcdefghijklmnopqrstuvwxyzABCD".split('').map.with_index { |l,i| [(i/10.0)+1,l] }]
  FLOAT_TO_LETTER_MAP[0.0] = '0'
  FLOAT_TO_LETTER_MAP[1.0] = '1'
  FLOAT_TO_LETTER_MAP[2.0] = '2'
  FLOAT_TO_LETTER_MAP[3.0] = '3'
  FLOAT_TO_LETTER_MAP[4.0] = '4'
  
  LETTER_TO_FLOAT_MAP = FLOAT_TO_LETTER_MAP.invert
  
  # NB: The Belgian calcualtor has some inputs 10..40 rather than 1..4
  # so adjust here. Also do maximum value check.
  def convert_letters_to_float(array)
    array.map.with_index do |entry,i|
      original = (LETTER_TO_FLOAT_MAP[entry].to_f || entry.to_f)
      type = ModelStructure.instance.types[i]
      if type == nil
        0
      elsif type < 10
        [original, type].min
      else
        [original * 10, type].min
      end
    end
  end
  
  CONTROL = (4..55).to_a.map { |r| "control_h#{r}"  } + (4..32).to_a.map { |r| "control_af#{r}" }
  
  def set_choices(code)
    choices = code.split('')
    choices = convert_letters_to_float(choices)
    set_array(CONTROL,choices)
    choices
  end
  
end

if __FILE__ == $0
  g = Belgium2050ModelResult.new

  tests = 100
  t = Time.now
  a = []
  tests.times do
    c = Belgium2050ModelResult::CONTROL.map { rand(4)+1 }.join
    a << r = g.calculate_pathway(c)
    p r[:sankey]
  end
  te = Time.now - t
  puts "#{te/tests} seconds per run"
  puts "#{tests/te} runs per second"
end
